chiselFrontend/src/main/scala/chisel3/core/Module.scala

```diff
@@ -157,91 +229,160 @@ extends HasId {
 
-    /** Recursively suggests names to supported "container" classes
-      * Arbitrary nestings of supported classes are allowed so long as the
-      * innermost element is of type HasId
-      * Currently supported:
-      *   - Iterable
-      *   - Option
-      * (Note that Map is Iterable[Tuple2[_,_]] and thus excluded)
-      */
-    def nameRecursively(prefix: String, nameMe: Any): Unit =
-      nameMe match {
-        case (id: HasId) => id.suggestName(prefix)
-        case Some(elt) => nameRecursively(prefix, elt)
-        case (iter: Iterable[_]) if iter.hasDefiniteSize =>
-          for ((elt, i) <- iter.zipWithIndex) {
-            nameRecursively(s"${prefix}_${i}", elt)
-          }
-        case _ => // Do nothing
-      }
-    /** Scala generates names like chisel3$util$Queue$$ram for private vals
-      * This extracts the part after $$ for names like this and leaves names
-      * without $$ unchanged
-      */
-    def cleanName(name: String): String = name.split("""\$\$""").lastOption.getOrElse(name)
-    for (m <- getPublicFields(classOf[Module])) {
-      nameRecursively(cleanName(m.getName), m.invoke(this))
-    }
+  // IO for this Module. At the Scala level (pre-FIRRTL transformations),
+  // connections in and out of a Module may only go through `io` elements.
+  def io: Record
 
-    // For Module instances we haven't named, suggest the name of the Module
-    _ids foreach {
-      case m: Module => m.suggestName(m.desiredName)
-      case _ =>
+  // Allow access to bindings from the compatibility package
+  protected def _ioPortBound() = _ports contains io
+
+  protected override def nameIds(rootClass: Class[_]): HashMap[HasId, String] = {
+    val names = super.nameIds(rootClass)
+
+    // Allow IO naming without reflection
+    names.put(io, "io")
+    names.put(clock, "clock")
+    names.put(reset, "reset")
+
+    names
+  }
+
+  private[core] override def generateComponent(): Component = {
+    _autoWrapPorts()  // pre-IO(...) compatibility hack
+
+    // Restrict IO to just io, clock, and reset
+    require(io != null, "Module must have io")
+    require(_ports contains io, "Module must have io wrapped in IO(...)")
+    require((_ports contains clock) && (_ports contains reset), "Internal error, module did not have clock or reset as IO")
+    require(_ports.size == 3, "Module must only have io, clock, and reset as IO")
+
+    super.generateComponent()
+  }
+
+  private[core] override def initializeInParent() {
+    // Don't generate source info referencing parents inside a module, since this interferes with
+    // module de-duplication in FIRRTL emission.
+    implicit val sourceInfo = UnlocatableSourceInfo
+
+    pushCommand(DefInvalid(sourceInfo, io.ref))
+
+    override_clock match {
+      case Some(override_clock) => clock := override_clock
+      case _ => clock := Builder.forcedClock
```

Comment:
Ditto, crash as would be the case if you create a Reg where you can't resolve forcedClock.